jetty-core/jetty-siwe/src/main/java/org/eclipse/jetty/security/siwe/internal/SignInWithEthereumParser.java

```diff
@@ -0,0 +1,65 @@
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.security.siwe.internal;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class SignInWithEthereumParser
+{
+    private static final String SCHEME_PATTERN = "[a-zA-Z][a-zA-Z0-9+\\-.]*";
+    private static final String DOMAIN_PATTERN = "(?:[a-zA-Z0-9\\-._~%]+@)?[a-zA-Z0-9\\-._~%]+(?:\\:[0-9]+)?";
+    private static final String ADDRESS_PATTERN = "0x[0-9a-fA-F]{40}";
+    private static final String STATEMENT_PATTERN = "[^\\n]*";
+    private static final String URI_PATTERN = "[^\\n]+";
+    private static final String VERSION_PATTERN = "[0-9]+";
+    private static final String CHAIN_ID_PATTERN = "[0-9]+";
+    private static final String NONCE_PATTERN = "[a-zA-Z0-9]{8}";
+    private static final String DATE_TIME_PATTERN = "\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(?:\\.\\d+)?(?:Z|[+-]\\d{2}:\\d{2})?";
+    private static final String REQUEST_ID_PATTERN = "[^\\n]*";
+    private static final String RESOURCE_PATTERN = "- " + URI_PATTERN;
+    private static final String RESOURCES_PATTERN = "(?:\n" + RESOURCE_PATTERN + ")*";
+    private static final Pattern SIGN_IN_WITH_ETHEREUM_PATTERN = Pattern.compile(
+        "^(?:(?<scheme>" + SCHEME_PATTERN + ")://)?(?<domain>" + DOMAIN_PATTERN + ") wants you to sign in with your Ethereum account:\n" +
+            "(?<address>" + ADDRESS_PATTERN + ")\n\n" +
+            "(?<statement>" + STATEMENT_PATTERN + ")?\n\n" +
+            "URI: (?<uri>" + URI_PATTERN + ")\n" +
+            "Version: (?<version>" + VERSION_PATTERN + ")\n" +
+            "Chain ID: (?<chainId>" + CHAIN_ID_PATTERN + ")\n" +
+            "Nonce: (?<nonce>" + NONCE_PATTERN + ")\n" +
+            "Issued At: (?<issuedAt>" + DATE_TIME_PATTERN + ")" +
+            "(?:\nExpiration Time: (?<expirationTime>" + DATE_TIME_PATTERN + "))?" +
+            "(?:\nNot Before: (?<notBefore>" + DATE_TIME_PATTERN + "))?" +
+            "(?:\nRequest ID: (?<requestId>" + REQUEST_ID_PATTERN + "))?" +
+            "(?:\nResources:(?<resources>" + RESOURCES_PATTERN + "))?$",
+        Pattern.DOTALL
+    );
+
+    private SignInWithEthereumParser()
+    {
+    }
+
+    public static SignInWithEthereumToken parse(String message)
+    {
+        Matcher matcher = SIGN_IN_WITH_ETHEREUM_PATTERN.matcher(message);
+        if (!matcher.matches())
+            return null;
+
+        return new SignInWithEthereumToken(matcher.group("scheme"), matcher.group("domain"),
+            matcher.group("address"), matcher.group("statement"), matcher.group("uri"),
+            matcher.group("version"), matcher.group("chainId"), matcher.group("nonce"),
+            matcher.group("issuedAt"), matcher.group("expirationTime"), matcher.group("notBefore"),
+            matcher.group("requestId"), matcher.group("resources"));
+    }
```

Comment:
Another static method only class.... move them all to a single `Etherium` or `EtheriumUtil` class

Reply:
I merged `EthereumUtil` and `EthereumSignatureVerifier`, but I would rather keep the parser separate.

It will just get messy forcing them all into one class. And we might make the parser non-static and stateful in the future.

Reply:
Then move these statics to the `SignInWithEthereumToken` class (whose name I still do not like)

Reply:
Ok I moved it to `SignInWithEthereumToken`.